### PR TITLE
Support file objects (streams, sockets etc.)

### LINKED ISSUE #657: File objects for Python dump/load
#599 added support for dumping and loading from streams in the C code. We want to add this functionality to the Python api also. This can largely follow the example set in the kastore code.

### PR DESCRIPTION
**Pull request by benjeffery**
Closes #657 

WIP! Using PR for testing on WIN/OSX.

### CODE REVIEW

**Comment by AdminBot-tskit** [comment]
📖 Docs for this PR can be previewed [here](https://tskit-dev.github.io/tskit-docs/909/)

**Comment by benjeffery** [comment]
There seems to be an issue with codecov not showing files, so haven't checked coverage yet. Will look into it.

File: python/tests/conftest.py [hunk 1]
```diff
@@ -1,4 +1,33 @@
...
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+Configuration and fixtures for pytest. Only put test-suite wide fixtures in here. Module
+specific fixtures should live in their modules.
```
Comment (jeromekelleher) [inline python/tests/conftest.py:24]: This is great! Can you just put in either a quick example of how to use these fixtures, or perhaps a link to the documentation from pytest for how this this is done?

**Review by jeromekelleher — APPROVED** [review]
LGTM! I haven't pored over the details of the tricky stuff, but I assume this is derived from what we did in kastore, so I don't need to?

File: python/tests/conftest.py (hunk 1 above)
Comment (benjeffery) [reply python/tests/conftest.py:24]: https://github.com/tskit-dev/tskit/pull/909/commits/c30dde114a74293c4ea7c956304dbdb08ce46780

**Comment by benjeffery** [comment]
> I assume this is derived from what we did in kastore, so I don't need to?

Yep, a lot of this was copied over.

File: python/tests/conftest.py [hunk 2]
```diff
@@ -53,9 +72,11 @@ def simple_ts():
     return msprime.simulate(2, random_seed=42)
 
 
-# A tree sequence with data in all fields
 @fixture(scope="session")
 def ts():
```
Comment (jeromekelleher) [inline python/tests/conftest.py:76]: This is great, but I worry a bit about attaching meaning to the (very commonly used) variable ``ts`` here. I can see this being confusing for people. What if we adopt a convention like having a suffix or prefix of ``fixture`` for these parameters? That might make things a bit more obvious.

File: python/tests/conftest.py [hunk 3]
```diff
@@ -100,5 +121,8 @@ def ts():
 
 
 @fixture(scope="session")
-def multi_ts():
-    return [msprime.simulate(i, random_seed=42) for i in range(2, 100, 5)]
+def multi_ts(ts):
```
Comment (jeromekelleher) [inline python/tests/conftest.py:124]: This is a bit vague, as I'd imagine we'll want all sorts of different fixtures like this. What about ``replicate_ts_fixture`` which will return a list of msprime replicate sims?

File: python/tests/conftest.py (hunk 2 above)
Comment (benjeffery) [reply python/tests/conftest.py]: Yeah, this kind of magic can be really confusing if you aren't aware what it going on. It seemed a bit much to put a suffix on, but when I think about the reader of the code it really does help. I've added them in https://github.com/tskit-dev/tskit/pull/909/commits/b0b566b26220b9a9214b378c3163b5586b341814

File: python/tests/conftest.py (hunk 3 above)
Comment (benjeffery) [reply python/tests/conftest.py]: Fixed in https://github.com/tskit-dev/tskit/pull/909/commits/b0b566b26220b9a9214b378c3163b5586b341814

File: python/tests/conftest.py (hunk 2 above)
Comment (jeromekelleher) [reply python/tests/conftest.py]: I agree, it feels a bit wordy in the code, but I think it'll help people who aren't used to using pytest.

**Comment by jeromekelleher** [comment]
LGTM, merge away whenever!
